_posts/2020-07-29-parallelized-sampling/parallelized-sampling.Rmd

```diff
@@ -0,0 +1,213 @@
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+As part of our recent work to support weighted sampling of Spark dataframes in `sparklyr`, we embarked on a journey searching for algorithms that can perform weighted sampling, especially sampling without replacement, in efficient and scalable ways within a distributed cluster-computing framework such as Apache Spark.
+
+In the interest of brevity, "weighted sampling without replacement" shall be shortened into `SWoR` for the remainder of this blog post.
+
+In the following sections, we will explain and illustrate what `SWoR` means probability-wise, briefly outline some alternative solutions we have considered but were not completely satisfied with, and then deep-dive into exponential variates, a simple mathematical construct that made the ideal solution for this problem possible.
+
+## How it all started?
+
+The journey started from a [Github issue](https://github.com/sparklyr/sparklyr/issues/2592) inquiring about the possibility of supporting the equivalent of `dplyr::sample_frac(..., weight = <weight_column>)` for Spark dataframes in `sparklyr`. For example,
+
+```{r}
+dplyr::sample_frac(mtcars, 0.25, weight = gear, replace = FALSE)
+```
+
+will randomly select one-fourth of all rows from a R dataframe named "mtcars" without replacement, using `mtcars$gear` as weights. We were unable to find any function implementing the weighted versions of `dplyr::sample_frac` among [Spark SQL built-in functions](https://spark.apache.org/docs/3.0.0/api/sql/index.html) in Spark 3.0 or in earlier versions, which means a future version of `sparklyr` will need to run its own weighted sampling algorithm to support such use cases.
+
+## What exactly is `SWoR`?
+
+Readers already having a crystal-clear mental picture of what `SWoR` entails should probably skip most of this section. The key take-away here is given $N$ rows $r_1, \dotsc, r_N$ and their weights $w_1, \dotsc, w_N$ and a desired sample size $n$, `SWoR` is conceptually equivalent to a $n$-step process of selecting 1 out of $n - j + 1$ remaining rows in the $j$-th step for $j \in \{1, \dotsc, n\}$, with each remaining row's likelihood of getting selected being linearly proportional to its weight in any of the steps, i.e.,
+
+```
+samples := {}
+population := {r[1], ..., r[N]}
+
+for j = 1 to n
+  select r[x] from population with probability
+    (w[x] / TotalWeight(population))
+  samples := samples + {r[x]}
+  population := population - {r[x]}
+```
+
+Therefore the probability of `SWoR` selecting a given sequence of $n$ rows $(r_1, \dotsc, r_n)$ is simply $\prod\limits_{j = 1}^{n} \left( {w_j} \middle/ {\sum\limits_{k = j}^{N}{w_k}} \right)$.
+
+Notice the outcome of a `SWoR` process is in fact order-significant, which is why in this post it will always be represented as an ordered tuple of elements.
+
+`SWoR` can also be explained in term of dart-throwing. For example, let's say the size of our sample space is 5:
```

Comment:
The math formula here simply describes the probability distribution of possible SWoR outcomes.

Even though it is straightforward, it definitely warrants a mentioning -- because the structure of this blog post is such that in the very next section, a slightly more complicated process is presented, and it will generate another seemingly different probability distribution which actually fits exactly the same description as the one mentioned in this section. So, the same formula must appear in both places.